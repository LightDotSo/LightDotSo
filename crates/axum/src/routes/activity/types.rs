// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::activity;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Activity root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Activity {
    /// The id of the activity.
    id: String,
    /// The entity id of the activity.
    entity: String,
    /// The operation type of the activity.
    operation: String,
    /// The timestamp of the activity.
    timestamp: String,
    /// The actor wallet address of the activity.
    address: Option<String>,
}

/// Implement From<activity::Data> for Activity.
impl From<activity::Data> for Activity {
    fn from(activity: activity::Data) -> Self {
        Self {
            id: activity.id,
            entity: activity.entity.to_string(),
            operation: activity.operation.to_string(),
            timestamp: activity.timestamp.to_rfc3339(),
            address: activity.wallet_address.map(|addr| addr.to_string()),
        }
    }
}
