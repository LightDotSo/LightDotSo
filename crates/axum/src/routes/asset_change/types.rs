// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::routes::{interpretation_action::types::InterpretationAction, token::types::Token};
use lightdotso_prisma::asset_change;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// AssetChange root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct AssetChange {
    /// The id of the asset change.
    id: String,
    /// The address of the asset change.
    address: String,
    /// The amount of the asset change.
    amount: i64,
    /// The before amount of the asset change.
    before_amount: Option<i64>,
    /// The after amount of the asset change.
    after_amount: Option<i64>,
    /// The action id of the asset change.
    action: Option<InterpretationAction>,
    /// The token id of the asset change.
    token: Option<Token>,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<asset_change::Data> for AssetChange.
impl From<asset_change::Data> for AssetChange {
    fn from(asset_change: asset_change::Data) -> Self {
        Self {
            id: asset_change.id,
            address: asset_change.address,
            amount: asset_change.amount.parse().unwrap_or_default(),
            before_amount: asset_change.before_amount.and_then(|s| s.parse::<i64>().ok()),
            after_amount: asset_change.after_amount.and_then(|s| s.parse::<i64>().ok()),
            action: asset_change.interpretation_action.and_then(|maybe_action| {
                maybe_action.map(|action| InterpretationAction::from(*action))
            }),
            token: asset_change
                .token
                .and_then(|maybe_token| maybe_token.map(|token| Token::from(*token))),
        }
    }
}
