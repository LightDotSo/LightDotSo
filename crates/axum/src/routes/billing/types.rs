// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::billing;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Billing root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Billing {
    /// The id of the billing.
    id: String,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing::Data> for Billing.
impl From<billing::Data> for Billing {
    fn from(billing: billing::Data) -> Self {
        Self { id: billing.id }
    }
}
