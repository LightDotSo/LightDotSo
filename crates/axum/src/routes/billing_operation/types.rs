// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lightdotso_prisma::billing_operation;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// BillingOperation root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct BillingOperation {
    /// The id of the billing operation.
    id: String,
    /// The status of the billing operation.
    status: String,
    /// The balance USD of the billing operation.
    balance_usd: f64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing_operation::Data> for BillingOperation.
impl From<billing_operation::Data> for BillingOperation {
    fn from(billing_operation: billing_operation::Data) -> Self {
        Self {
            id: billing_operation.id,
            status: billing_operation.status.to_string(),
            balance_usd: billing_operation.balance_usd,
        }
    }
}
