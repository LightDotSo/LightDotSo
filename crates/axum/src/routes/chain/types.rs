// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::chain;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Chain root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Chain {
    /// The id of the protocol group.
    id: i64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<chain::Data> for Chain.
impl From<chain::Data> for Chain {
    fn from(chain: chain::Data) -> Self {
        Self { id: chain.id }
    }
}
