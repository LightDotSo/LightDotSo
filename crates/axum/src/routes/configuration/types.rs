// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::{configuration, owner};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Configuration root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct Configuration {
    /// The id of the configuration.
    id: String,
    /// The address of the configuration.
    address: String,
    /// The image hash of the configuration.
    image_hash: String,
    /// The checkpoint of the configuration.
    checkpoint: i64,
    /// The threshold of the configuration.
    threshold: i64,
    /// The owners of the configuration.
    owners: Vec<ConfigurationOwner>,
}

/// Configuration Owner.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct ConfigurationOwner {
    /// The id of the owner.
    id: String,
    /// The index of the owner.
    index: i32,
    /// The address of the owner.
    address: String,
    /// The weight of the owner.
    weight: i64,
}

/// Implement From<configuration::Data> for Configuration.
impl From<configuration::Data> for Configuration {
    fn from(configuration: configuration::Data) -> Self {
        Self {
            id: configuration.id.to_string(),
            address: configuration.address.to_string(),
            image_hash: configuration.image_hash.to_string(),
            checkpoint: configuration.checkpoint,
            threshold: configuration.threshold,
            owners: configuration.owners.map_or(Vec::new(), |owners| {
                owners.into_iter().map(ConfigurationOwner::from).collect()
            }),
        }
    }
}

/// Implement From<owner::Data> for Owner.
impl From<owner::Data> for ConfigurationOwner {
    fn from(owner: owner::Data) -> Self {
        Self {
            id: owner.id.to_string(),
            address: owner.address.to_string(),
            index: owner.index,
            weight: owner.weight,
        }
    }
}
