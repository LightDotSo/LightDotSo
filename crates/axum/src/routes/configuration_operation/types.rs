// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lightdotso_prisma::configuration_operation;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// ConfigurationOperation root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct ConfigurationOperation {
    /// The id of the configuration operation.
    pub id: String,
    /// The image hash of the configuration operation.
    pub image_hash: String,
    /// The checkpoint of the configuration operation.
    pub checkpoint: i64,
    /// The threshold of the configuration operation.
    pub threshold: i64,
    /// The status of the configuration operation.
    pub status: String,
    /// The timestamp of the user operation.
    pub created_at: String,
    /// The timestamp updated of the user operation.
    pub updated_at: String,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<configuration_operation::Data> for ConfigurationOperation.
impl From<configuration_operation::Data> for ConfigurationOperation {
    fn from(configuration_operation: configuration_operation::Data) -> Self {
        Self {
            id: configuration_operation.id,
            image_hash: configuration_operation.image_hash,
            checkpoint: configuration_operation.checkpoint,
            threshold: configuration_operation.threshold,
            status: configuration_operation.status.to_string(),
            created_at: configuration_operation.created_at.to_rfc3339(),
            updated_at: configuration_operation.updated_at.to_rfc3339(),
        }
    }
}
