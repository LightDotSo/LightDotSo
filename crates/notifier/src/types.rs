// Copyright 2023-2024 Light
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lazy_static::lazy_static;
use lightdotso_prisma::{ActivityEntity, ActivityOperation};
use serde_json::Value;
use std::collections::BTreeMap;
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter, EnumString, IntoStaticStr};

// This is a list of all platforms
#[derive(Clone, Debug, EnumString, EnumIter, IntoStaticStr, Display)]
pub enum Platform {
    #[strum(serialize = "WEB")]
    Web,
}

// This is a list of all user operations
#[derive(Clone, Debug, EnumString, EnumIter, IntoStaticStr, Display)]
pub enum UserOnlyOperation {
    #[strum(serialize = "INVITE_CODE_ACCEPTED")]
    InviteCodeAccepted,
}

// Utility function to match the activity entity/operation combination
pub fn match_user_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<UserOnlyOperation> {
    match (entity, operation) {
        (ActivityEntity::InviteCode, ActivityOperation::Update) => {
            // Check if the key of log is status is `USED`
            if log["status"].as_str() == Some("USED") {
                Some(UserOnlyOperation::InviteCodeAccepted)
            } else {
                None
            }
        }
        _ => None,
    }
}

// This is a list of all wallet operations
#[derive(Clone, Debug, EnumString, EnumIter, IntoStaticStr, Display)]
pub enum WalletOnlyOperation {
    #[strum(serialize = "USER_OPERATION_CREATED")]
    UserOperationCreated,
    #[strum(serialize = "USER_OPERATION_EXECUTED")]
    UserOperationExecuted,
    #[strum(serialize = "TRANSACTION_WITH_USER_OPERATION_EXECUTED")]
    TransactionWithUserOperationExecuted,
}

// Utility function to match the activity entity/operation combination
pub fn match_wallet_only_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<WalletOnlyOperation> {
    match (entity, operation) {
        (ActivityEntity::UserOperation, ActivityOperation::Create) => {
            Some(WalletOnlyOperation::UserOperationCreated)
        }
        (ActivityEntity::UserOperation, ActivityOperation::Update) => {
            if log["status"].as_str() == Some("EXECUTED") ||
                log["status"].as_str() == Some("REVERTED")
            {
                Some(WalletOnlyOperation::TransactionWithUserOperationExecuted)
            } else {
                None
            }
        }
        (ActivityEntity::Transaction, ActivityOperation::Update) => None,
        _ => None,
    }
}

// Combination utility function to match the activity entity/operation combination
pub fn match_notification_with_activity(
    entity: &ActivityEntity,
    operation: &ActivityOperation,
    log: &Value,
) -> Option<Operation> {
    if let Some(res) = match_user_only_notification_with_activity(entity, operation, log) {
        return Some(Operation::UserOnly(res));
    }

    match_wallet_only_notification_with_activity(entity, operation, log).map(Operation::WalletOnly)
}

// This is a list of all operations
#[derive(Clone, Debug, Display)]
pub enum Operation {
    UserOnly(UserOnlyOperation),
    WalletOnly(WalletOnlyOperation),
}

lazy_static! {
    // This is a list of all operations
    pub static ref OPERATIONS: Vec<Operation> = {
        let mut operations = Vec::new();

        for operation in UserOnlyOperation::iter() {
            operations.push(Operation::UserOnly(operation));
        }

        for operation in WalletOnlyOperation::iter() {
            operations.push(Operation::WalletOnly(operation));
        }

        operations
    };
}

// For each operation, we define a notification type which is all combinations of operation and
// platform
#[derive(Clone, Debug)]
pub enum Notification {
    UserOnly(Platform, UserOnlyOperation),
    WalletOnly(Platform, WalletOnlyOperation),
}

// Implement Display for Notification
impl std::fmt::Display for Notification {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Notification::UserOnly(platform, operation) => {
                write!(f, "{}-{}", platform, operation)
            }
            Notification::WalletOnly(platform, operation) => {
                write!(f, "{}-{}", platform, operation)
            }
        }
    }
}

lazy_static! {
    // This is a list of all notification types
    pub static ref NOTIFICATION: Vec<Notification> = {
        let mut notification_types = Vec::new();

        for operation in UserOnlyOperation::iter() {
            for platform in Platform::iter() {
                notification_types.push(Notification::UserOnly(platform, operation.clone()));
            }
        }

        for operation in WalletOnlyOperation::iter() {
            for platform in Platform::iter() {
                notification_types.push(Notification::WalletOnly(platform, operation.clone()));
            }
        }

        notification_types
    };
}

lazy_static! {
    // This is a list of all user notification types
    pub static ref USER_NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for notif in NOTIFICATION.iter() {
            if let Notification::UserOnly(_, operation) = notif {
                let is_default_enabled = match operation {
                    UserOnlyOperation::InviteCodeAccepted => true,
                };
                map.insert(notif.to_string(), is_default_enabled);
            }
        }

        map
    };
}

lazy_static! {
    // This is a list of all wallet notification types
    pub static ref WALLET_NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for notif in NOTIFICATION.iter() {
            if let Notification::WalletOnly(_, operation) = notif {
                let is_default_enabled = match operation {
                    WalletOnlyOperation::UserOperationCreated => true,
                    WalletOnlyOperation::UserOperationExecuted => true,
                    WalletOnlyOperation::TransactionWithUserOperationExecuted => true,
                };
                map.insert(notif.to_string(), is_default_enabled);
            }
        }

        map
    };
}

lazy_static! {
    // This is a list of all notification default enabled
    pub static ref NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for (key, value) in &*USER_NOTIFICATION_DEFAULT_ENABLED {
            map.insert(key.clone(), *value);
        }

        for (key, value) in &*WALLET_NOTIFICATION_DEFAULT_ENABLED {
            map.insert(key.clone(), *value);
        }

        map
    };
}

lazy_static! {
    // This is a list of all user notification keys
    pub static ref USER_NOTIFICATION_KEYS: Vec<String> = {
        let mut user_notification_keys = Vec::new();

        for notif in NOTIFICATION.iter() {
            if let Notification::UserOnly(_, _) = notif {
                user_notification_keys.push(notif.to_string());
            }
        }

        user_notification_keys
    };
}

lazy_static! {
    // This is a list of all wallet notification keys
    pub static ref WALLET_NOTIFICATION_KEYS: Vec<String> = {
        let mut wallet_notification_keys = Vec::new();

        for notif in NOTIFICATION.iter() {
            if let Notification::WalletOnly(_, _) = notif {
                wallet_notification_keys.push(notif.to_string());
            }
        }

        wallet_notification_keys
    };
}

lazy_static! {
    // This is a list of all notification keys
    pub static ref NOTIFICATION_KEYS: Vec<String> = {
        let mut notification_keys = Vec::new();

        for notification in NOTIFICATION.iter() {
            notification_keys.push(notification.to_string());
        }

        notification_keys
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_notification_keys() {
        for key in USER_NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!("USER_NOTIFICATION_KEYS", format!("{:?}", *USER_NOTIFICATION_KEYS));
    }

    #[test]
    fn test_wallet_notification_keys() {
        for key in WALLET_NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!(
            "WALLET_NOTIFICATION_KEYS",
            format!("{:?}", *WALLET_NOTIFICATION_KEYS)
        );
    }

    #[test]
    fn test_notification_keys() {
        for key in NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!("NOTIFICATION_KEYS", format!("{:?}", *NOTIFICATION_KEYS));
    }

    #[test]
    fn test_operations() {
        for operation in OPERATIONS.iter() {
            println!("{}", operation);
        }
        insta::assert_snapshot!("OPERATIONS", format!("{:?}", *OPERATIONS));
    }

    #[test]
    fn test_notification() {
        for notification in NOTIFICATION.iter() {
            println!("{}", notification);
        }
        insta::assert_snapshot!("NOTIFICATION", format!("{:?}", *NOTIFICATION));
    }

    #[test]
    fn test_notification_default_enabled() {
        insta::assert_snapshot!(
            "NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *NOTIFICATION_DEFAULT_ENABLED)
        );
    }

    #[test]
    fn test_user_notification_default_enabled() {
        insta::assert_snapshot!(
            "USER_NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *USER_NOTIFICATION_DEFAULT_ENABLED)
        );
    }

    #[test]
    fn test_wallet_notification_default_enabled() {
        insta::assert_snapshot!(
            "WALLET_NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *WALLET_NOTIFICATION_DEFAULT_ENABLED)
        );
    }
}
