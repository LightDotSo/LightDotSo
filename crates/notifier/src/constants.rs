// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::types::{
    Notification, NotificationOperation, Platform, UserOnlyNotification, WalletOnlyNotification,
};
use lazy_static::lazy_static;
use std::collections::BTreeMap;
use strum::IntoEnumIterator;

// -----------------------------------------------------------------------------
// Constants
// -----------------------------------------------------------------------------

lazy_static! {
    // This is a list of all notification types
    pub static ref NOTIFICATIONS: Vec<Notification> = {
        let mut notifications= Vec::new();

        for operation in UserOnlyNotification::iter() {
            for platform in Platform::iter() {
                notifications.push(Notification::UserOnly(platform, operation.clone()));
            }
        }

        for operation in WalletOnlyNotification::iter() {
            for platform in Platform::iter() {
                notifications.push(Notification::WalletOnly(platform, operation.clone()));
            }
        }

        notifications
    };
}

lazy_static! {
    // This is a list of all notification operations
    pub static ref NOTIFICATION_OPERATIONS: Vec<NotificationOperation> = {
        let mut notification_operations = Vec::new();

        for notification_operation in UserOnlyNotification::iter() {
            notification_operations.push(NotificationOperation::UserOnly(notification_operation));
        }

        for notification_operation in WalletOnlyNotification::iter() {
            notification_operations.push(
                NotificationOperation::WalletOnly(notification_operation),
            );
        }

        notification_operations
    };
}

lazy_static! {
    // This is a list of all user notification types
    pub static ref USER_NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for notif in NOTIFICATIONS.iter() {
            if let Notification::UserOnly(_, operation) = notif {
                let is_default_enabled = match operation {
                    UserOnlyNotification::InviteCodeAccepted => true,
                };
                map.insert(notif.to_string(), is_default_enabled);
            }
        }

        map
    };
}

lazy_static! {
    // This is a list of all wallet notification types
    pub static ref WALLET_NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for notif in NOTIFICATIONS.iter() {
            if let Notification::WalletOnly(_, operation) = notif {
                let is_default_enabled = match operation {
                    WalletOnlyNotification::UserOperationCreated => true,
                    WalletOnlyNotification::UserOperationExecuted => true,
                    WalletOnlyNotification::TransactionWithUserOperationExecuted => true,
                };
                map.insert(notif.to_string(), is_default_enabled);
            }
        }

        map
    };
}

lazy_static! {
    // This is a list of all notification default enabled
    pub static ref NOTIFICATION_DEFAULT_ENABLED: BTreeMap<String, bool> = {
        let mut map = BTreeMap::new();

        for (key, value) in &*USER_NOTIFICATION_DEFAULT_ENABLED {
            map.insert(key.clone(), *value);
        }

        for (key, value) in &*WALLET_NOTIFICATION_DEFAULT_ENABLED {
            map.insert(key.clone(), *value);
        }

        map
    };
}

lazy_static! {
    // This is a list of all user notification keys
    pub static ref USER_NOTIFICATION_KEYS: Vec<String> = {
        let mut user_notification_keys = Vec::new();

        for notif in NOTIFICATIONS.iter() {
            if let Notification::UserOnly(_, _) = notif {
                user_notification_keys.push(notif.to_string());
            }
        }

        user_notification_keys
    };
}

lazy_static! {
    // This is a list of all wallet notification keys
    pub static ref WALLET_NOTIFICATION_KEYS: Vec<String> = {
        let mut wallet_notification_keys = Vec::new();

        for notif in NOTIFICATIONS.iter() {
            if let Notification::WalletOnly(_, _) = notif {
                wallet_notification_keys.push(notif.to_string());
            }
        }

        wallet_notification_keys
    };
}

lazy_static! {
    // This is a list of all notification keys
    pub static ref NOTIFICATION_KEYS: Vec<String> = {
        let mut notification_keys = Vec::new();

        for notification in NOTIFICATIONS.iter() {
            notification_keys.push(notification.to_string());
        }

        notification_keys
    };
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_notification_keys() {
        for key in USER_NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!("USER_NOTIFICATION_KEYS", format!("{:?}", *USER_NOTIFICATION_KEYS));
    }

    #[test]
    fn test_wallet_notification_keys() {
        for key in WALLET_NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!(
            "WALLET_NOTIFICATION_KEYS",
            format!("{:?}", *WALLET_NOTIFICATION_KEYS)
        );
    }

    #[test]
    fn test_notification_keys() {
        for key in NOTIFICATION_KEYS.iter() {
            println!("{}", key);
        }
        insta::assert_snapshot!("NOTIFICATION_KEYS", format!("{:?}", *NOTIFICATION_KEYS));
    }

    #[test]
    fn test_notifications() {
        for notification in NOTIFICATIONS.iter() {
            println!("{}", notification);
        }
        insta::assert_snapshot!("NOTIFICATIONS", format!("{:?}", *NOTIFICATIONS));
    }

    #[test]
    fn test_notification_operations() {
        for operation in NOTIFICATION_OPERATIONS.iter() {
            println!("{}", operation);
        }
        insta::assert_snapshot!(
            "NOTIFICATION_OPERATIONS",
            format!("{:?}", *NOTIFICATION_OPERATIONS)
        );
    }

    #[test]
    fn test_notification_default_enabled() {
        insta::assert_snapshot!(
            "NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *NOTIFICATION_DEFAULT_ENABLED)
        );
    }

    #[test]
    fn test_user_notification_default_enabled() {
        insta::assert_snapshot!(
            "USER_NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *USER_NOTIFICATION_DEFAULT_ENABLED)
        );
    }

    #[test]
    fn test_wallet_notification_default_enabled() {
        insta::assert_snapshot!(
            "WALLET_NOTIFICATION_DEFAULT_ENABLED",
            format!("{:?}", *WALLET_NOTIFICATION_DEFAULT_ENABLED)
        );
    }
}
