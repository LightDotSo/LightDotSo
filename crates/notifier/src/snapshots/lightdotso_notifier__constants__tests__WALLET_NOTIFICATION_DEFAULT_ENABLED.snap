---
source: crates/notifier/src/constants.rs
expression: "format!(\"{:?}\", *WALLET_NOTIFICATION_DEFAULT_ENABLED)"
---
{"WEB:::TRANSACTION_WITH_USER_OPERATION_EXECUTED": true, "WEB:::USER_OPERATION_CREATED": true, "WEB:::USER_OPERATION_EXECUTED": true}
